crates/extensions/c8y_mapper_ext/src/availability/tests.rs

```diff
@@ -414,24 +413,50 @@ async fn timer_send(timer: &mut FakeServerBox<TimerStart, TimerComplete>, event:
 }
 
 struct TestHandler {
-    pub mqtt_box: SimpleMessageBox<PublishMessage, MqttMessage>,
+    pub mqtt_box: SimpleMessageBox<MqttMessage, MqttMessage>,
     pub timer_box: FakeServerBox<TimerStart, TimerComplete>,
 }
 
 async fn spawn_availability_actor(config: AvailabilityConfig) -> TestHandler {
-    let mut mqtt_builder: SimpleMessageBoxBuilder<PublishMessage, MqttMessage> =
+    // create two separate Mqtt message boxes to circumvent "can't borrow the same object as mutable more than one time"
+    // use the message boxes because they implement `MessageSource` and `MessageSink` (receivers and senders alone do
+    // not) and then take out their halves not given to the actor and recombine them
+    //
+    // we need separate mqtt in and mqtt out because on the in-side CumulocityMapperActor attaches a `ChannelFilter`
+    // which we use to only get messages we care about, and on the out-side we want to output directly to the MQTT actor
+    // because on CumulocityMapperActor side the halves that send to AvailabilityActor and receive from
+    // AvailabilityActor aren't separate, so one can block the other
+    //
+    // TODO: find a cleaner and more permanent solution
```

Comment:
One solution, is to fully disconnect the  availability actor from the c8y actor and let it consume its input directly from an MQTT actor.

Reply:
> One solution, is to fully disconnect the availability actor from the c8y actor and let it consume its input directly from an MQTT actor.

Routing the MQTT messages from the availability actor to the MQTT actor via the intermediary C8y actor was an intentional design choice to avoid a race condition that we encountered when a direct connection was implemented in the first drop of that availability feature.

With the old design where the availability actor and c8y actor were connected directly to the MQTT actor independently, when a new device was getting registered, the c8y actor would map and publish the device creation message and the availability actor would publish the availability message as well independently. This was causing a race condition where sometimes the availability message was reaching the cloud first, before the device creation message, which messes up the device metadata.

So, we needed a mechanism to make sure that the availability messages for an entity are pushed to the cloud only after its respective registration message is already processed and pushed. By routing the MQTT messages from the availability actor to the C8y actor, we could ensure this sequentiality. So, if we're breaking that design, we must find another solution to the race condition as well.